### DISCUSSION TITLE
`composeStory` + custom projectDecorations not merged with `setProjectAnnotations`?

### DISCUSSION THREAD

**Discussion by csantos1113** [storybookjs/storybook#25766] [Help]
### Summary

Is it possible to merge my per-test projectDecorations with whatever was set via `setProjectAnnotations`?
I was expecting this to be the case, similar to how stories can override pieces of the global configuration; but that doesn't seem to be the case for `setProjectAnnotations`

### Additional information

For example: in normal storybook lands:

`preview.tsx`
```tsx
import type { Preview } from '@storybook/react';

const preview: Preview = {
  decorators: [
    (Story, { parameters }) => {
      return (
        <>
          <div>
            my decorator!!
            <code>{JSON.stringify(parameters.user, null, 2)}</code>
          </div>
          <Story />
        </>
      );
    },
  ],
  parameters: {
    user: {
      location: 'main',
      role: 'VIEWER',
    },
  },
};

export default preview;
```

in my story I simply have

```tsx
import type { Meta, StoryObj } from '@storybook/react';
import { Button } from './Button';

const meta = {
  title: 'Example/Button',
  component: Button
} satisfies Meta<typeof Button>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Primary: Story = {
  args: {
    primary: true,
    label: 'Button',
  },
};

export const AdminButton: Story = {
  ...Primary,
  parameters: {
    user: {
      role: 'ADMIN'
    }
  },
}
```

`AdminButton` overrides `role` to ADMIN, but it still shows the other value coming from `preview.tsx` ("location: main")

<img width="300" alt="Screenshot 2024-01-25 at 4 49 01 PM" src="https://github.com/storybookjs/storybook/assets/9648559/43385fe7-493e-442b-aaf8-d0b976a08fc2">
<img width="300" alt="Screenshot 2024-01-25 at 4 49 05 PM" src="https://github.com/storybookjs/storybook/assets/9648559/cf6cd083-98de-472b-8d37-26a34a05c781">

---

But it doesn't seem to be the same case when using `setProjectAnnotations` and `composeStory`

```tsx
import * as globalStorybookConfig from './.storybook/preview.tsx';

import { composeStory, setProjectAnnotations } from '@storybook/react';

setProjectAnnotations(globalStorybookConfig);

import Meta, { Primary } from './Button.stories';

const AdminButton = composeStory(Primary, Meta, {
  parameters: {
    user: {
      role: 'ADMIN'
    }
  },
});
```

The entire `decorators` list is not executed

---

Shouldn't it behave as in "normal storybook"?

### Create a reproduction

https://stackblitz.com/edit/github-fkycya

**Comment by csantos1113** [comment]
It seems once I provide a custom project annotation, the entire "global project annotation" is ignored

https://github.com/storybookjs/storybook/blob/7ecf05e5f5b27872c478134f7345a87f9beb63cc/code/lib/preview-api/src/modules/store/csf/testing-utils/index.ts#L37

I wonder if this could be a _merge-deep_ similar to how "normal storybook" does it

**Comment by Marklb** [comment]
I see your point, but I think it is working as intended, based on the docs description. The docs describe it as a way to override the global config, which does not sound like it is intended to merge.

The docs description of that feature: https://storybook.js.org/docs/writing-tests/stories-in-unit-tests#override-story-properties
> By default, the `setProjectAnnotations` function injects into your existing tests any global configuration you've defined in your Storybook instance (i.e., parameters, decorators in the preview.js|ts file). Nevertheless, this may cause unforeseen side effects for tests that are not intended to use these global configurations. To avoid this, you can override the global configurations by extending either the `composeStory` or `composeStories` functions to provide test-specific configurations.


I think the intention of composing a story is to take an already configured story and compose it into a single object that can be used to render that story. If I understand correctly, you are trying to basically use a composed story as a base to extend into a different story.

For now, maybe you could try one of the following:

**Option 1**: Compose the configs, yourself.
```ts
import * as globalStorybookConfig from './.storybook/preview.tsx';

import { composeStory, setProjectAnnotations, composeConfigs } from '@storybook/react';

setProjectAnnotations(globalStorybookConfig);

import Meta, { Primary } from './Button.stories';

const AdminButton = composeStory(Primary, Meta,
  composeConfigs([
    globalStorybookConfig,
    {
      parameters: {
        user: {
          role: 'ADMIN'
        }
      },
    }
  ])
);
```

**Option 2**: Define a new story to compose. *This may not be simple, if you use the old format that defines the story as a function.*
```ts
import * as globalStorybookConfig from './.storybook/preview.tsx';

import { composeStory, setProjectAnnotations } from '@storybook/react';

setProjectAnnotations(globalStorybookConfig);

import Meta, { Primary } from './Button.stories';

const AdminStory = {
  ...Primary,
  parameters: {
    ...Primary.parameters,
    user: {
      ...Primary.parameters.user,
      role: 'ADMIN'
    }
  }
}

const AdminButton = composeStory(AdminStory, Meta);
```

**Reply by csantos1113** [reply]
Hey @Marklb thanks for the quick response. It indeed works as mentioned in the docs

> Nevertheless, this may cause unforeseen side effects for tests that are not intended to use these global configurations. To avoid this, you can override the global configurations by extending either the composeStory or composeStories functions to provide test-specific configurations.

But that's the thing, I wonder if it should work as "merge" instead as "override".

Anything that is in the `preview.ts` is useful and applied for all stories (in live storybook), so wouldn't this concern "this may cause unforeseen side effects for tests that are not intended to use these global configurations" be also true for live stories? and in that case we do can expand or modify these props easily when defining the story object, i.e:

```tsx
export const Primary: Story = {};

// highlight-next-line
export const AdminButton: Story = {
  // this is from preview.ts but can be extended here
  decorators: [],
  // this is from preview.ts but can be extended here
  loaders: [],
  // this is from preview.ts but can be extended here
  parameters: {
    user: {
      role: 'ADMIN'
    }
  },
  // except globalTypes? https://github.com/storybookjs/storybook/issues/13634
}
```

But we can't achieve the same using `composeStory`.

So, to this point "_If I understand correctly, you are trying to basically use a composed story as a base to extend into a different story._": I'd say, not really. I want to have a base story and compose 2 different renderable stories for my unit tests.

I think I get the idea of _"composing a story is to take an already configured story and compose it into a single object that can be used to render that story"_ but I can't stop thinking that it should be more streamline to use inherit/extend that global config in the tests 😅
